tools/src/main/java/com/nedap/archie/rmobjectvalidator/RMObjectValidationUtil.java

```diff
@@ -26,8 +26,8 @@
     public static String getParentObservationTerm(CAttribute attribute) {
         String result = "";
         CObject parent = attribute.getParent();
-        while (result.equals("") && parent != null) {
-            CAttribute attributeParent = parent.getParent();
+        CAttribute attributeParent = parent.getParent();
+        while (result.equals("") && parent != null && attributeParent != null) {
             if (attributeParent != null) {
```

Comment:
How about
```java
public static String getParentObservationTerm(CAttribute attribute) {
        String result = "";
        CObject parent = attribute.getParent();
        CAttribute attributeParent = parent.getParent();
        while (result.equals("") && parent != null && attributeParent != null) {
            parent = attributeParent.getParent();
            if (parent != null && parent.getRmTypeName().equals("OBSERVATION")) {
                ArchetypeTerm parentTerm = parent.getTerm();
                if (parentTerm != null) {
                    result = parentTerm.getText();
                }
            }
            attributeParent = parent == null ? null : parent.getParent();
        }
        return result;
    }
```

Reply:
Just happened to see this randomly. I had a look at the code in Archie. This function only exists to enable an error message to be output by the function `isObservationEmpty()`.  I am surprised to see that the variable `archetype` is not retained from the call `validate()`, which then obviates this kind of complicated code that has to walk up the parent chain to get to the archetype and then get to the terminology, so as to be able to convert the id-code on a node to its human-readable form. 

I might be missing something, but if not, I think this will simplify this class quite a bit.

Reply:
There can be several observations inside one archetype if it is an OPT. This means this still need to walk back to parents to find the correct observation, and if the archetype is passed it will not be of use.

Reply:
You mean the `archetype` argument here could be an OPT structure? I see in the current tests it is a source (differential) archetype (or maybe it's flattened, depending on how much work `ADLParser.parse()` does).

Reply:
Yes, this should be an OperationalTemplate, otherwise it will stop validating archetype constraints at the first use_archetype and perhaps even use_node statement - note the lack of an archetype repository parameter to retrieve other archetypes. It will continue to validate the RM constraints of course.
The validator does not make a different since both source, flat or OPT are all just archetypes. But in hindsight requiring an OPT would probably have been better here. Can probably be changed, but that would be another pull request than this one.

Should be also OPT in tests, but it does not matter if the source archetypes do not have a parent and do not use use_node or use_archetype - then the OPT or flat form is the same as the source form.

Reply:
And everywhere where the Archetype class is used, it can be an OPT structure unless explicitly documented and checked that it is in source form, because an OPT is simply a subclass of Archetype.